.github/workflows/compilation.yml

```diff
@@ -2,106 +2,184 @@ name: CI
 
 on:
   push:
+    paths-ignore:
+      - '**.md'
+      - '**.rst'
   pull_request:
+    paths-ignore:
+      - '**.md'
+      - '**.rst'
   repository_dispatch:
     types: [run_build, run_release]
   workflow_dispatch: {}
+env:
+  HOMEBREW_NO_ANALYTICS: 1
+  HOMEBREW_NO_AUTO_UPDATE: 1
+  HOMEBREW_NO_INSTALL_UPGRADE: 1
+  HOMEBREW_NO_INSTALL_CLEANUP: 1
+  DEBIAN_FRONTEND: noninteractive
+  TZ: Japan
 
 jobs:
   build:
-    runs-on: ${{ matrix.target.machine }}
     strategy:
       matrix:
-        target:
-          - machine: macos-latest
+        os: 
+          - runs-on: macos-latest
             architecture: arm64
             shell: bash
-            packageManager: brew
-          - machine: macos-13
+            container: null
+            container-options: " "
+            macos-package-manager: brew
+          - runs-on: macos-13
             architecture: x86_64
             shell: bash
-            packageManager: brew
-          - machine: macos-latest
+            container: null
+            container-options: " "
+            macos-package-manager: brew
+          - runs-on: macos-latest
             architecture: arm64
             shell: bash
-            packageManager: macport
-          - machine: ubuntu-latest
+            container: null
+            container-options: " "
+            macos-package-manager: port
+          - runs-on: ubuntu-latest
             architecture: x86_64
             shell: bash
-            packageManager: default
-          - machine: ubuntu-22.04-arm
+            container: "ubuntu:20.04"
+            container-options: "--user 0"
+            macos-package-manager: null
+          - runs-on: ubuntu-24.04-arm
             architecture: arm64
             shell: bash
-            packageManager: default
-          - machine: ubuntu-24.04-arm
-            architecture: arm64
-            shell: bash
-            packageManager: default
-          - machine: windows-latest
+            container: "ubuntu:20.04"
+            container-options: "--user 0"
+            macos-package-manager: null
+          - runs-on: windows-latest
             architecture: x86_64
             shell: msys2
-            packageManager: default
+            container: null
+            container-options: " "
+            macos-package-manager: null
       fail-fast: false
+    runs-on: ${{ matrix.os.runs-on }}
     defaults:
      run:
-      shell: ${{ matrix.target.shell }} {0}
+      shell: ${{ matrix.os.shell }} {0}
+    container:
+      image: ${{ matrix.os.container }}
+      options: ${{ matrix.os.container-options }}
+    timeout-minutes: 240
```

Comment:
why are you setting `timeout`? I think it doesn't make sense in our scenario

Reply:
This is in case the build hangs (e.g. stuck at prompt or stuck in infinite retry loop) it will allow to notice earlier. Github default runners have a maximum timeout of 6 hours, so I reduced it to 4 hours.